graph-viz/app/treeCut/TreeCut.java

```diff
@@ -0,0 +1,238 @@
+package treeCut;
+
+import clustering.PointCluster;
+import models.Cluster;
+import models.Edge;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+
+/**
+ * Implementation of tree cut algorithm
+ */
+public class TreeCut {
+
+    /**
+     * Main process of tree cut algorithm.
+     *
+     * @param pointCluster    hierarchical structure from HGC algorithm
+     * @param lowerLongitude  lowerLongitude of user screen
+     * @param upperLongitude  upperLongitude of user screen
+     * @param lowerLatitude   lowerLatitude of user screen
+     * @param upperLatitude   upperLatitude of user screen
+     * @param zoom            zoom level of user screen
+     * @param edges           edge set to be returned
+     * @param externalEdgeSet edge set with only one node inside screen
+     * @param externalCluster outside cluster corresponding to edge set with only one node inside screen
+     * @param internalCluster inside screen clusters
+     */
+    public void treeCut(PointCluster pointCluster, double lowerLongitude,
+                        double upperLongitude, double lowerLatitude,
+                        double upperLatitude, int zoom,
+                        HashMap<Edge, Integer> edges, HashSet<Edge> externalEdgeSet,
+                        HashSet<Cluster> externalCluster, HashSet<Cluster> internalCluster) {
+        // The final ancestor cluster to be mapped to for each external cluster
+        HashMap<Cluster, Cluster> externalClusterMap = new HashMap<>();
+        // Current ancestor points to the list of its descendant external clusters
+        HashMap<Cluster, ArrayList<Cluster>> externalHierarchy = new HashMap<>();
+        // Current ancestor points of internal clusters
+        HashSet<Cluster> internalHierarchy = new HashSet<>();
+        // initialize the hierarchy
+        addExternalClusterHierarchy(externalCluster, externalHierarchy, externalClusterMap);
+        addInternalClusterHierarchy(internalCluster, internalHierarchy);
+        // TODO change breaking clause
+        while (externalHierarchy.size() != 0) {
+            // The ancestors has conflict
+            ArrayList<Cluster> removeAncestors = new ArrayList<>();
+            // Find clusters has common ancestor with internal clusters at this level
+            for (Map.Entry<Cluster, ArrayList<Cluster>> clusterEntry : externalHierarchy.entrySet()) {
```

Comment:
Instead of using removeAncestor to remove unwanted items, use stream.map to delete